src/middleware.ts

```diff
@@ -137,30 +176,45 @@ export const persist = <S extends State>(
 ) => (set: SetState<S>, get: GetState<S>, api: StoreApi<S>): S => {
   const {
     name,
-    storage = typeof localStorage !== 'undefined'
-      ? localStorage
-      : {
-          getItem: () => null,
-          setItem: () => {},
-        },
+    storage: getStorage = () => localStorage,
     serialize = JSON.stringify,
     deserialize = JSON.parse,
     blacklist,
     whitelist,
-    postRehydrationMiddleware,
+    onRehydrateStorage,
+    version,
   } = options || {}
 
+  let storage: StateStorage | undefined
+
+  try {
+    storage = getStorage()
+  } catch (e) {
+    // prevent error if the storage is not defined (e.g. when server side rendering a page)
+  }
+
+  if (!storage) return config(set, get, api)
+
   const setItem = async () => {
-    const state = { ...get() }
+    let state = { ...get() }
+
     if (whitelist) {
-      Object.keys(state).forEach(
-        (key) => !whitelist.includes(key) && delete state[key]
-      )
+      Object.keys(state).forEach((key) => {
+        !whitelist.includes(key) && delete state[key]
+      })
     }
     if (blacklist) {
       blacklist.forEach((key) => delete state[key])
     }
-    storage.setItem(name, await serialize(state))
+
+    if (version) {
+      state = {
+        ...state,
+        __version: version,
```

Comment:
This is dangerous, it could conflict with the same property name used in the state.
Can we make it safe?

```js
const data = {
  state,
  version,
}
storage.setItem(name, serialize(data))
```

It's another breaking change...

Or something like:
```js
storage.setItem(name, serialize(state))
storage.setItem(`${name}.version`, serialize(version))
```

Reply:
```js
storage.setItem(name, serialize(state))
storage.setItem(`${name}.version`, serialize(version))
```
That's a good idea 👍

Reply:
After thinking about it, I think it's better to keep everything in one place.